### PR TITLE
fix(Core/Raid): Eyes in Naxxramas

### PR DESCRIPTION
**Pull request by AsunaFrostwyrm**
## CHANGES PROPOSED:
- Added the GUID-catch for the eyes in Naxxramas in the instance-script
- Added activation of the eyes when an endboss of a quarter is dead

## ISSUES ADDRESSED:
- Closes https://github.com/azerothcore/azerothcore-wotlk/issues/2818
- Closes https://github.com/azerothcore/azerothcore-wotlk/issues/2909

## TESTS PERFORMED:
- Built successfully on Win10 64 and Debian 9
- Killed Maexxna and went back to the entrance, the eye is lighted up
- Resetted instances, killed Four Horsemans this time, eye of the quarter is lighted up aswell
- Killed all endbosses, all four eyes were lighted up like this:

image

## HOW TO TEST THE CHANGES:

1. Kill an endboss of a quarter in naxxramas
2. Go back to the entrance and look at the eye which is placed on the ramp in the direction to the quarter you cleared
3. Did it light up?

## KNOWN ISSUES AND TODO LIST:

- None

## Target branch(es):
- [x] Master

 
## How to test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here in the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

### CODE REVIEW

**Comment by BarbzYHOOL** [comment]
cool little detail added

**Comment by LukasVolgger** [comment]
Renamed to our commit template. Good job!

**Comment by AsunaFrostwyrm** [comment]
Fixed the wrong indentation now. Hopefully it is okay now!

File: src/server/scripts/Northrend/Naxxramas/instance_naxxramas.cpp [hunk 1]
```diff
@@ -338,7 +346,7 @@ class instance_naxxramas : public InstanceMapScript
                     _kelthuzadgateGUID = pGo->GetGUID();
                     if (GetBossState(BOSS_SAPPHIRON) == DONE && _speakTimer==0)
                         pGo->SetGoState(GO_STATE_ACTIVE);
-               	   break;
+               	    break;
```
Comment (Nefertumm) [inline src/server/scripts/Northrend/Naxxramas/instance_naxxramas.cpp]: space

File: src/server/scripts/Northrend/Naxxramas/naxxramas.h [hunk 2]
```diff
@@ -84,6 +84,11 @@ enum NXGOs
     GO_MAEXXNA_PORTAL               = 181575, //Maexxna portal
     GO_HORSEMAN_PORTAL              = 181578, //Four Horseman portal
 
+    GO_DEATHKNIGHT_EYE_PORTAL		    = 181210,
+    GO_PLAGUE_EYE_PORTAL			      = 181211,
+    GO_SPIDER_EYE_PORTAL			      = 181212,
+    GO_ABOM_EYE_PORTAL				      = 181213,
```
Comment (Nefertumm) [inline src/server/scripts/Northrend/Naxxramas/naxxramas.h]: Align those equals and we're good to go :)

File: src/server/scripts/Northrend/Naxxramas/instance_naxxramas.cpp (hunk 1 above)
Comment (AsunaFrostwyrm) [reply src/server/scripts/Northrend/Naxxramas/instance_naxxramas.cpp]: This space is intended. While I was correcting my fault indentation I also checked the indentation of the cases above and saw that this "break" statement is missing a space in front of it.

Here is a screenshot of the original source in the current rev:

indentation

Thought I could fix it within this PR, but I can also remove it if you're saying that I'm wrong tho :)

Comment (Nefertumm) [reply src/server/scripts/Northrend/Naxxramas/instance_naxxramas.cpp]: Oh oh, then it's good :)

File: src/server/scripts/Northrend/Naxxramas/naxxramas.h (hunk 2 above)
Comment (AsunaFrostwyrm) [reply src/server/scripts/Northrend/Naxxramas/naxxramas.h]: Should be finally fixed now :)

**Comment by Si1ker** [comment]
@AsunaFrostwyrm This will be working only for the first time after you reset the realm the eye for the 4H will not be working, probably it is not being saved.

**Comment by AsunaFrostwyrm** [comment]
Can confirm that. Anyhow I really don't know why this eye doesn't light up again after a restart. The encounter is correctly saved in the database after killing the four horseman. Additionally, the portal which appears at the horsemans, which you can use to teleport yourself back to the beginning of Naxxramas sets itself to phase 1 (as intended). For this phasechange the script checks for

```C++
if (GetBossState(BOSS_HORSEMAN) == DONE)
```

Thats what it exactly does for the eye at the beginning of naxxramas. I don't understand why it doesn't work to be honest..

**Comment by AsunaFrostwyrm** [comment]
The latest commit also closes this issue: https://github.com/azerothcore/azerothcore-wotlk/issues/2909

Finally I found out, what caused the problem of the instance not loading the correct bossstate when it loads the states from the DB. 

In "instance_naxxramas.cpp" the bossstates are set with some restrictions, like the four horseman in this case. With this if-clause:
```C++
if (bossId == BOSS_HORSEMAN)
```
and the following code, the bossstate of BOSS_HORSEMAN will not be set to "DONE" until 4 horseman are killed, because of these codelines:
```C++
if (state == DONE)
                {
                    _horsemanTimer++;
                    _horsemanKilled++;
                    if (_horsemanKilled < 4) {
                        return false;
                    }

                    // All horsemans are killed
                    if (Creature* cr = instance->GetCreature(_blaumeuxGUID))
                        cr->CastSpell(cr, 59450, true); // credit
                }
```
When the core unloads the instance after a given time (or because of a restart), it also unloads the counter of "_horsemanKilled" and thats why it doesn't set the bossstate to "DONE" after reloading the states from the DB.

What I did now is to insert another bool-variable to check, if the bossstate is set to "DONE" in the overridden void Load(const char* in) function. If the bossstate is loaded as "DONE", this variable sets to "true" and the if-clause from the beginning will not be true anymore, because I added another comparison into the if-clause:
```C++
if (bossId == BOSS_HORSEMAN && _horsemanLoadDoneState == false)
```

**Comment by BarbzYHOOL** [comment]
> bossstate

stop talking in Parseltongue :snake:

**Comment by Nefertumm** [comment]
Tested, working really nice :)

**Comment by BarbzYHOOL** [comment]
@AsunaFrostwyrm thank you! merged

thank you @Nefertumm for testing and reviewing

@Si1ker thank you too for testing

@LukasVolgger thank you for laughing at my joke

File: src/server/scripts/Northrend/Naxxramas/instance_naxxramas.cpp [hunk 3]
```diff
@@ -497,7 +527,7 @@ class instance_naxxramas : public InstanceMapScript
             }
 
             // Horseman handling
-            if (bossId == BOSS_HORSEMAN)
+            if (bossId == BOSS_HORSEMAN && _horsemanLoadDoneState == false)
```
Comment (FrancescoBorzi) [inline src/server/scripts/Northrend/Naxxramas/instance_naxxramas.cpp:530]: can be just `bossId == BOSS_HORSEMAN &&  !_horsemanLoadDoneState`

**Pull request merged**